cpp/src/arrow/engine/expression.cc

```diff
@@ -0,0 +1,416 @@
+    if (ARROW_PREDICT_FALSE(cond)) {         \
+      return Status::ErrorType(__VA_ARGS__); \
+    }                                        \
+  } while (false)
+
+#define ERROR_IF(cond, ...) ERROR_IF_TYPE(cond, Invalid, __VA_ARGS__)
+
+//
+// Expr
+//
+
+std::string Expr::kind_name() const {
+  switch (kind_) {
+    case ExprKind::SCALAR_LITERAL:
+      return "scalar";
+    case ExprKind::FIELD_REFERENCE:
+      return "field_ref";
+    case ExprKind::COMPARE_OP:
+      return "compare_op";
+    case ExprKind::EMPTY_REL:
+      return "empty_rel";
+    case ExprKind::SCAN_REL:
+      return "scan_rel";
+    case ExprKind::PROJECTION_REL:
+      return "projection_rel";
+    case ExprKind::FILTER_REL:
+      return "filter_rel";
+  }
+
+  return "unknown expr";
+}
+
+struct ExprEqualityVisitor {
+  bool operator()(const ScalarExpr& rhs) const {
+    auto lhs_scalar = internal::checked_cast<const ScalarExpr&>(lhs);
+    return lhs_scalar.scalar()->Equals(*rhs.scalar());
+  }
+
+  bool operator()(const FieldRefExpr& rhs) const {
+    auto lhs_field = internal::checked_cast<const FieldRefExpr&>(lhs);
+    return lhs_field.index() == rhs.index() &&
+           lhs_field.operand()->Equals(*rhs.operand());
+  }
+
+  template <typename E>
+  enable_if_compare_expr<E, bool> operator()(const E& rhs) const {
+    auto lhs_cmp = internal::checked_cast<const E&>(lhs);
+    return (lhs_cmp.left_operand()->Equals(rhs.left_operand()) &&
+            lhs_cmp.right_operand()->Equals(rhs.right_operand())) ||
+           (lhs_cmp.left_operand()->Equals(rhs.right_operand()) &&
+            lhs_cmp.left_operand()->Equals(rhs.right_operand()));
+  }
+
+  bool operator()(const EmptyRelExpr& rhs) const {
+    auto lhs_empty = internal::checked_cast<const EmptyRelExpr&>(lhs);
+    return lhs_empty.schema()->Equals(rhs.schema());
+  }
+
+  bool operator()(const ScanRelExpr& rhs) const {
+    auto lhs_scan = internal::checked_cast<const ScanRelExpr&>(lhs);
+    // Performs a pointer equality on Table/Dataset
+    return lhs_scan.input() == rhs.input();
+  }
+
+  bool operator()(const ProjectionRelExpr& rhs) const {
+    auto lhs_proj = internal::checked_cast<const ProjectionRelExpr&>(lhs);
+
+    const auto& lhs_exprs = lhs_proj.expressions();
+    const auto& rhs_exprs = rhs.expressions();
+    if (lhs_exprs.size() != rhs_exprs.size()) {
+      return false;
+    }
+
+    for (size_t i = 0; i < lhs_exprs.size(); i++) {
+      if (!lhs_exprs[i]->Equals(rhs_exprs[i])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  bool operator()(const Expr&) const { return false; }
+
+  static bool Visit(const Expr& lhs, const Expr& rhs) {
+    return VisitExpr(rhs, ExprEqualityVisitor{lhs});
+  }
+
+  const Expr& lhs;
+};
+
+bool Expr::Equals(const Expr& other) const {
+  if (this == &other) {
+    return true;
+  }
+
+  if (kind() != other.kind() || type() != other.type()) {
+    return false;
+  }
+
+  return ExprEqualityVisitor::Visit(*this, other);
```

Comment:
If every expr type has a unique name then this could be implemented using an `EqualsInternal` virtual